KaleidoscopeLang/Parser.swift

```diff
@@ -75,10 +75,10 @@ private let valueExpression: ValueExpressionParser = fix { valueExpression in
 private let prototypeArgs = %(.Character("(")) *> many(identifier) <* %(.Character(")"))
 
 /// prototype ::= Identifier prototypeArgs
-private let prototype = PrototypeExpression.init <^> ( lift(pair) <*> identifier <*> prototypeArgs )
+private let prototype = curry(PrototypeExpression.init) <^> identifier <*> prototypeArgs
 
 /// definition ::= "def" prototype expression
-private let definition: TopLevelExpressionParser = FunctionExpression.init <^> ( %(Token.Def) *> lift(pair) <*> prototype <*> valueExpression )
+private let definition: TopLevelExpressionParser = %(Token.Def) *> (curry(FunctionExpression.init) <^> prototype <*> valueExpression)
```

Comment:
This is the only strange case. Because of the left-to-right evaluation order of applicative operators, we have to reorder the token before the applied function, or else both it and the applied function would be discarded (and we wouldn’t be able to typecheck).

We require the parentheses for much the same reason: `*>` would otherwise be applied to a function, which it doesn’t know how to deal with. It’s possible that this is a case for `lift`; I haven’t really thought about it that much.

Overall I’d like to discourage the use of `lift`, however :blush: